File: 2024/12-19/python/Part2.py
import os

def count_ways(design, patterns, memo):
    if design == "":
        return 1

    if design in memo:
        return memo[design]

    count = 0
    for pattern in patterns:
        if design.startswith(pattern):
            remaining = design[len(pattern):]
            count += count_ways(remaining, patterns, memo)

    memo[design] = count
    return count

def main():
    # Input
    with open("e:/source/advent2024/12-19/towels.txt", "r") as file:
        lines = file.readlines()

    patterns = sorted([pattern.strip() for pattern in lines[0].strip().split(',')], key=len)
    designs = lines[2:]

    # Calculate the number of ways for each design
    memo = {}
    total_ways = 0

    for design in designs:
        design = design.strip()
        ways = count_ways(design, patterns, memo)
        total_ways += ways

    print(f"Total Ways: {total_ways}")

if __name__ == "__main__":
    main()

File: 2024/12-19/go/Part2.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "sort"
    "strings"
)

// CountWays calculates the number of ways to create the design using patterns.
func CountWays(design string, patterns []string, memo map[string]int64) int64 {
    if design == "" {
        return 1
    }

    if val, exists := memo[design]; exists {
        return val
    }

    var count int64
    for _, pattern := range patterns {
        if strings.HasPrefix(design, pattern) {
            remaining := design[len(pattern):]
            count += CountWays(remaining, patterns, memo)
        }
    }

    memo[design] = count
    return count
}

func main() {
    // Open the file
    file, err := os.Open("e:/source/advent2024/12-19/towels.txt")
    if err != nil {
        fmt.Println("Error opening file:", err)
        return
    }
    defer file.Close()

    // Read the file
    scanner := bufio.NewScanner(file)
    var lines []string
    for scanner.Scan() {
        lines = append(lines, scanner.Text())
    }

    if err := scanner.Err(); err != nil {
        fmt.Println("Error reading file:", err)
        return
    }

    // Extract patterns and designs
    patterns := strings.Split(lines[0], ",")
    for i := range patterns {
        patterns[i] = strings.TrimSpace(patterns[i])
    }
    sort.SliceStable(patterns, func(i, j int) bool {
        return len(patterns[i]) < len(patterns[j])
    })

    designs := lines[2:]

    // Calculate the number of ways for each design
    memo := make(map[string]int64)
    var totalWays int64

    for _, design := range designs {
        ways := CountWays(design, patterns, memo)
        totalWays += ways
    }

    fmt.Printf("Total Ways: %d\n", totalWays)
}